webpack/scenes/ContentViews/Publish/PublishContentViewWizard.js

```diff
@@ -0,0 +1,132 @@
+import {
+  selectEnvironmentPaths,
+  selectEnvironmentPathsStatus,
+} from '../components/EnvironmentPaths/EnvironmentPathSelectors';
+
+const PublishContentViewWizard = ({
+  details, show, setIsOpen, currentStep, setCurrentStep,
+}) => {
+  const { name, id: cvId, version_count: versionCount } = details;
+  const [description, setDescription] = useState('');
+  const [checkedItems, setCheckedItems] = useState([]);
+  const [promote, setPromote] = useState(false);
+  const [forcePromote, setForcePromote] = useState([]);
+  const dispatch = useDispatch();
+  const environmentPathResponse = useSelector(selectEnvironmentPaths);
+  const environmentPathStatus = useSelector(selectEnvironmentPathsStatus);
+  const environmentPathLoading = environmentPathStatus === STATUS.PENDING;
+
+  const steps = [
+    {
+      id: 1,
+      name: 'Publish',
+      component: <CVPublishForm
+        description={description}
+        setDescription={setDescription}
+        details={details}
+        checkedItems={checkedItems}
+        setCheckedItems={setCheckedItems}
+        promote={promote}
+        setPromote={setPromote}
+        forcePromote={forcePromote}
+      />,
+    },
+    {
+      id: 2, name: 'Review', component: <CVPublishReview details={details} checkedItems={checkedItems} />, nextButtonText: 'Finish',
+    },
+    {
+      id: 3,
+      name: 'Finish',
+      component: <CVPublishFinish
+        description={description}
+        setDescription={setDescription}
+        checkedItems={checkedItems}
+        setCheckedItems={setCheckedItems}
+        forcePromote={forcePromote}
+        cvId={cvId}
+        versionCount={versionCount}
+        show={show}
+        setIsOpen={setIsOpen}
+        currentStep={currentStep}
+        setCurrentStep={setCurrentStep}
+      />,
+      isFinishedStep: true,
+    },
+  ];
+
+  useEffect(
+    () => {
+      dispatch(getEnvironmentPaths());
+      setCurrentStep(1);
+    },
+    [],
+  );
+
+  useEffect(() => {
+    if (!environmentPathLoading) {
+      const { results } = environmentPathResponse || {};
+      let forcedPromoteList = [];
+      results.forEach((result) => {
+        const { environments: path } = result;
+        // Library and first env in any path can never be forced so we can ignore those
+        const forcedPathCandidates = path.slice(2);
+        forcedPathCandidates.forEach((env) => {
+          if (forcedPromoteList.filter(item => item.id === env.prior.id).length) {
+            // Add any env to force whose prior is forced.
+            forcedPromoteList = [...forcedPromoteList, env];
+          } else if (!(checkedItems.filter(item => item.id === env.prior.id).length) &&
+            checkedItems.includes(env)) {
+            // Find and add first out of order environment
+            forcedPromoteList = [...forcedPromoteList, env];
+          }
+        });
+      });
+      setForcePromote(forcedPromoteList.filter(item => (checkedItems.includes(item))));
+    }
```

Comment:
This works fine but I still think it's hard to read and grok what's going on.  Are there ways you can refactor it to make it easier to comprehend?  (I actually like the recursion idea and think it makes it simpler in some ways, for example you don't have to have a special case for the first two paths.)